R/cnd.R

```diff
@@ -435,6 +435,87 @@ validate_signal_args <- function(msg, type, call, env = parent.frame()) {
   }
 }
 
+#' Muffle a condition
+#'
+#' Unlike [exiting()] handlers, [calling()] handlers must be explicit
+#' that they have handled a condition to stop it from propagating to
+#' other handlers. Use `cnd_muffle()` within a calling handler (or as
+#' a calling handler, see examples) to prevent any other handlers from
+#' being called for that condition.
+#'
+#'
+#' @section Mufflable conditions:
+#'
+#' Not all conditions are signalled with a muffling restart.
+#' `cnd_muffle()` is compatible with:
+#'
+#' * `warning` and `message` conditions. In this case `cnd_muffle()`
+#'   is equivalent to [base::suppressMessages()] and
+#'   [base::suppressWarnings()].
+#'
+#' * Bare conditions signalled with `signal()` or [cnd_signal()].
+#'
+#' * Interrupts are sometimes signalled with a `resume` restart on
+#'   recent R versions.
+#'
+#' If you call `cnd_muffle()` with a condition that is not mufflable
+#' you will cause a new error to be signalled.
+#'
+#' * Errors are not mufflable since they are signalled in critical
+#'   situations where execution cannot continue safely.
+#'
+#' * Conditions captured with [base::tryCatch()], [with_handlers()] or
```

Comment:
I don't understand this

Reply:
I still don't understand. Can you give an example?

Reply:
ok jumping to a wrong restart is harder to trigger than I thought. It cannot happen when using only our signalling functions, the user would have to set up a muffle restart manually:

```r
impl <- function() {
  signal("", "my_cnd")
  "expected return value"
}

fn_correct <- function() {
  with_handlers(impl(), my_cnd = cnd_muffle)
}
fn_wrong <- function() {
  with_handlers(impl(), my_cnd = ~cnd_muffle(.x))
}

with_restarts(muffle = function() "wrong return value",
  fn_correct()
)
#> [1] "expected return value"

with_restarts(muffle = function() "wrong return value",
  fn_wrong()
)
#> [1] "wrong return value"
```

Maybe we could prefix the restart (e.g. `rlang_muffle`) to make this event even less likely.


On the other hand getting a "no restart" error is easy, especially with the formula shortcut interpreted as exiting by default:

```r
# This is fine:
with_handlers(warn(""), warning = cnd_muffle)

# These cause errors:
with_handlers(warn(""), warning = ~cnd_muffle(.x))
#> Error in invokeRestart("muffleWarning") :
#>   no 'restart' 'muffleWarning' found

tryCatch(warning(""), warning = cnd_muffle)
#> Error in invokeRestart("muffleWarning") :
#>   no 'restart' 'muffleWarning' found
```

Maybe we shouldn't use `as_function()` in `with_handlers()` and force users to be explicit with `exiting()` when they want to use formulas

Reply:
I think that current behaviour is ok - you can't muffle in an exiting handler.

Reply:
Sounds good. Is this PR now mergeable?